scala/compiler-jps/src/org/jetbrains/jps/incremental/scala/SbtBuilder.scala

```diff
@@ -31,8 +32,8 @@ import _root_.scala.collection.mutable
 class SbtBuilder extends ModuleLevelBuilder(BuilderCategory.TRANSLATOR) {
   override def getPresentableName = "Scala sbt builder"
 
-  override def buildStarted(context: CompileContext): Unit = {
-    if (isScalaProject(context) && !isDisabled(context)) {
+  override def chunkBuildStarted(context: CompileContext, chunk: ModuleChunk): Unit = {
+    if (isScalaProject(context) && !isDisabled(context) && hasScalaModules(context, chunk)) {
```

Comment:
I will analyze that tomorrow, thanks for pointing out.

Reply:
Ok, I checked that. So it actually fixes a bug. If in scala project you create a purely java module without scala sdk, it will error with `No scalac found to compile scala sources: ...` (from org/jetbrains/jps/incremental/scala/local/SbtCompiler.scala:23). So if zinc is to be used, it needs scala sdk as well. After this fix, the module would be compiled with regular java builder (no zinc involved).

Reply:
> If in scala project you create a purely java module without scala sdk, it will error with No scalac found to compile scala sources: ... (from org/jetbrains/jps/incremental/scala/local/SbtCompiler.scala:23). So if zinc is to be used, it needs scala sdk as well.
Hm... that doesn't look correct - it seems that it should be possible to compile Java-only modules, without Scala SDK attached. How does SBT compile Java-only modules?

> After this fix, the module would be compiled with regular java builder (no zinc involved).
But... that will prevent the to-JAR compilation. I also wonder whether incremental compilation will work in such a case, in both directions - we need to verify that.

(Speaking of incremental compilation - it supposed to be integrated with other languages / compilers / GUI forms in IDEA. What happens if we store the output classes in a JAR? It seems that we need to list specific use cases and test each of them to make sure that everything works.)

Reply:
> Hm... that doesn't look correct - it seems that it should be possible to compile Java-only modules, without Scala SDK attached. How does SBT compile Java-only modules?

Probably sbt is implemented differently to always pass those jars or something. I don't want to test again, I am pretty sure it was like that because I was testing it when I was writing the original comment. Try to create some scala project and add java module that does not depend on it and does not have scala sdk and try to build it with zinc. It should fail with the error I mentioned. After the fix it will be built without zinc.

> But... that will prevent the to-JAR compilation. I also wonder whether incremental compilation will work in such a case, in both directions - we need to verify that.

Yes, it prevents compilation to jar and incremental compilation, but aside from that it produces output, you can run apps from there and so on. If one wants zinc features, one must add scala sdk. 
Is there any other solution you were thinking about?

> (Speaking of incremental compilation - it supposed to be integrated with other languages / compilers / GUI forms in IDEA. What happens if we store the output classes in a JAR? It seems that we need to list specific use cases and test each of them to make sure that everything works.)

There are two areas of that. I am not sure which are you asking about.
 * Integrating compilation to jar with language / compiler 
Highly depends on how incremental compiler work. The example of such integration is in my zinc PR mentioned in first post. It involves appending new classes to jar, pruning invalidated classes, etc. We operate on a single jar instead of the file system. There are multiple tricks used there to keep it efficient.
* Integrating compile to jar with intellij
For the use-cases we have it is mostly building and running apps. Both handled in this PR. For running we had to add jars to classpath. Resources are still copied normally (not included in jar) so we have 2 classpath entries instead of 1.
The other  thing is running tests. AFAIR the logic in intellij for junit that we were used is just skipping jars explicitly when looking for tests. We have it solved by setting java property `junit4.search.4.tests.in.classpath=false` so that it falls back to different method of looking for tests. We were not looking at things like packaging/generating artifacts, gui forms. 
I also recall one more thing. If user is running an app but at the same time wants to edit and compile the code that the app uses it will fail, as jar will be locked. As a workaround I implemented a logic that when app is run, creates a copy of the compiled jar with `_runtime` suffix. Only the runtime jars are used for running. I check the timestamps to avoid overhead. It is quite performant even on windows. I hooked it into `RunConfigurationExtension#updateJavaParameters` for our custom run configurations that we are using for all apps. I am not sure how could it be implemented in general case.